labelbox/data/serialization/ndjson/label.py

```diff
@@ -45,40 +101,57 @@ def from_common(cls,
-                elif isinstance(annotation, NDSegments):
-                    annots.extend(
-                        NDSegments.to_common(annotation, annotation.name,
-                                             annotation.schema_id))
-                elif isinstance(annotation, NDDicomMasks):
-                    annots.append(NDDicomMasks.to_common(annotation))
-                elif isinstance(annotation, NDVideoMasks):
-                    annots.append(NDVideoMasks.to_common(annotation))
-                elif isinstance(annotation, NDObjectType.__args__):
-                    annots.append(NDObject.to_common(annotation))
-                elif isinstance(annotation, NDClassificationType.__args__):
-                    annots.extend(NDClassification.to_common(annotation))
-                elif isinstance(annotation,
+        for _, group in annotation_groups.items():
+            annotations = []
+            # first, we iterate through all the NDJSON objects and store the
+            # deserialized objects in the _AnnotationGroupTuple
+            # object *if* the object can be used in a relationship
+            for uuid, ndjson in group.ndjson_objects.items():
+                if isinstance(ndjson, NDDicomSegments):
+                    annotations.extend(
+                        NDDicomSegments.to_common(ndjson, ndjson.name,
+                                                  ndjson.schema_id))
+                elif isinstance(ndjson, NDSegments):
+                    annotations.extend(
+                        NDSegments.to_common(ndjson, ndjson.name,
+                                             ndjson.schema_id))
+                elif isinstance(ndjson, NDDicomMasks):
+                    annotations.append(NDDicomMasks.to_common(ndjson))
+                elif isinstance(ndjson, NDVideoMasks):
+                    annotations.append(NDVideoMasks.to_common(ndjson))
+                elif isinstance(ndjson, NDObjectType.__args__):
+                    annotation = NDObject.to_common(ndjson)
+                    annotations.append(annotation)
+                    group.annotations[uuid] = annotation
+                elif isinstance(ndjson, NDClassificationType.__args__):
+                    annotations.extend(NDClassification.to_common(ndjson))
+                elif isinstance(ndjson,
                                 (NDScalarMetric, NDConfusionMatrixMetric)):
-                    annots.append(NDMetricAnnotation.to_common(annotation))
+                    annotations.append(NDMetricAnnotation.to_common(ndjson))
                 else:
-                    raise TypeError(
-                        f"Unsupported annotation. {type(annotation)}")
-            yield Label(annotations=annots,
-                        data=self._infer_media_type(data_row, annots))
+                    raise TypeError(f"Unsupported annotation. {type(ndjson)}")
+
+            # after all the annotations have been discovered, we can now create
```

Comment:
Good point! So, if I understand your question correctly, for this to happen it would mean that source and/or target annotations from a relationship belong to a different DataRow and AFAIK, that's not a valid scenario, or is it actually possible?

Reply:
Yes, we only will have one annotation group per datarow.
If multiple labels were created, like:
```
    labels = [
        lb_types.Label(data=lb_types.ImageData(global_key=...),
                       annotations=[annotations_1]),
        lb_types.Label(data=lb_types.ImageData(global_key=...),
                       annotations=[annotations_2]),
        ...
    ]
```
then all these annotations would be grouped under a single key in the annotation group, as they all belong to the same datarow identified by the datarow-id or global-key.